File: src/surf/platforms/youtube.ts
/**
 * Relay Surf — YouTube adapter (server-only).
 *
 * No public unauthenticated YouTube *search* API exists, so:
 *   - search: site-scoped web indexing of youtube.com (honest — labeled
 *     "web-index", never claimed as native YouTube search).
 *   - extract: YouTube's public oEmbed + watch-page metadata for titles,
 *     channels, and descriptions of a specific video.
 *   - transcript: best-effort caption extraction from the public watch page;
 *     reported as unavailable when captions aren't accessible.
 */
import { relaiFetch } from "@/RelAI/web/http.server";
import { assertSafeUrl } from "../guard";
import type { PlatformAdapter, PlatformSearchOptions, SearchResult } from "../types";

function videoIdFromUrl(url: string): string | null {
  try {
    const u = new URL(url);
    if (u.hostname === "youtu.be") return u.pathname.split("/")[1] ?? null;
    if (u.hostname.includes("youtube.com")) {
      return u.searchParams.get("v");
    }
  } catch {
    /* not a URL */
  }
  return null;
}

export const youtubeAdapter: PlatformAdapter = {
  platform: "youtube",
  canHandle(url: string) {
    try {
      const host = new URL(url).hostname;
      return host === "youtube.com" || host.endsWith(".youtube.com") || host === "youtu.be";
    } catch {
      return false;
    }
  },
  async search(query, options: PlatformSearchOptions = {}) {
    const { relaiSearch } = await import("@/RelAI/web/search.server");
    const limit = options.limit ?? 8;
    const res = await relaiSearch(query, { site: "youtube.com", limit: Math.min(limit * 2, 20) });
    const results: SearchResult[] = res.hits
      .filter((h) => /youtube\.com\/watch|youtu\.be\//i.test(h.url) || /youtube\.com\/(@|c\/|channel|user)/i.test(h.url))
      .slice(0, limit)
      .map((h) => ({
        url: h.url,
        title: h.title,
        snippet: h.snippet,
        provider: "web-index",
        accessMode: "public",
        platform: "youtube",
      }));
    return {
      results,
      accessMode: "public",
      note:
        "YouTube has no public unauthenticated search API. Results come from public web indexing of youtube.com.",
    };
  },
  async extractPage(url) {
    const safeUrl = assertSafeUrl(url);
    const videoId = videoIdFromUrl(safeUrl);
    if (!videoId) return null;

    // oEmbed is a stable public metadata endpoint.
    const oembed = await relaiFetch(
      `https://www.youtube.com/oembed?url=${encodeURIComponent(`https://www.youtube.com/watch?v=${videoId}`)}&format=json`,
      { timeoutMs: 10_000, retries: 1, cache: true },
    );
    let meta: { title?: string; author_name?: string; author_url?: string } = {};
    if (oembed.ok) {
      try {
        meta = JSON.parse(oembed.text);
      } catch {
        /* keep empty meta */
      }
    }

    // Description + captions from the public watch page (best-effort).
    const watch = await relaiFetch(`https://www.youtube.com/watch?v=${videoId}`, {
      timeoutMs: 15_000,
      retries: 1,
      cache: true,
    });

    let description = "";
    if (watch.ok) {
      const m = watch.text.match(/"shortDescription":"([\s\S]*?)"/);
      if (m) description = m[1].replace(/\\n/g, "\n").slice(0, 2000);
    }

    // Transcript: check for accessible caption tracks. The captions endpoint
    // is only available when the uploader enables them — reported honestly.
    let transcript = "";
    if (watch.ok) {
      const trackRe = /"captionTracks":\[[^\]]*\]/;
      const tracks = watch.text.match(trackRe);
      if (tracks) {
        const baseUrl = tracks[0].match(/"baseUrl":"([^"]+)"/);
        if (baseUrl) {
          const capRes = await relaiFetch(baseUrl[1].replace(/\\u0026/g, "&"), {
            timeoutMs: 10_000,
            retries: 1,
            cache: true,
          });
          if (capRes.ok) {
            const xml = capRes.text;
            const lines = [...xml.matchAll(/<text[^>]*>([\s\S]*?)<\/text>/g)]
              .map((t) => t[1].replace(/&amp;/g, "&").replace(/&quot;/g, '"').replace(/&#39;/g, "'"))
              .slice(0, 400)
              .join(" ");
            if (lines.length > 40) transcript = lines;
          }
        }
      }
    }

    if (!meta.title && !watch.ok) return null;

    const text = [
      meta.title ? `# ${meta.title}` : `# Video ${videoId}`,
      meta.author_name ? `Channel: ${meta.author_name} (${meta.author_url ?? ""})` : "",
      ``,
      description ? `## Description\n${description}` : "",
      ``,
      transcript ? `## Transcript (public captions)\n${transcript}` : "## Transcript\n(no accessible public captions)",
    ]
      .filter(Boolean)
      .join("\n");

    return {
      url: safeUrl,
      title: meta.title ?? `Video ${videoId}`,
      text,
      links: [],
      headings: [{ level: 1, text: meta.title ?? `Video ${videoId}` }],
      metadata: {
        channel: meta.author_name ?? "",
        channelUrl: meta.author_url ?? "",
        videoId,
        transcript: transcript ? "available" : "unavailable",
      },
      truncated: transcript.length > 0,
      fetchedAt: new Date().toISOString(),
      status: 200,
      contentType: "text/html",
      accessMode: "public",
    };
  },
};

File: app/platforms/youtube.py
"""YouTube platform adapter — search videos, extract metadata, chapters, transcripts."""

from __future__ import annotations

import re
from typing import Optional
from urllib.parse import urlparse, parse_qs

import httpx

from .base import PlatformAdapter
from ..core.types import PageData, SearchResult, SearchAccessMode, SurfPlatform
from ..core.http import relai_fetch
from ..config import get_settings

# ── URL patterns ──────────────────────────────────────────────────────
_YT_HOSTS = {"youtube.com", "www.youtube.com", "m.youtube.com", "youtu.be"}


def _video_id(url: str) -> Optional[str]:
    """Extract video ID from any YouTube URL format."""
    parsed = urlparse(url)
    if parsed.hostname in ("youtu.be",):
        return parsed.path.strip("/").split("/")[0]
    if parsed.hostname in _YT_HOSTS:
        qs = parse_qs(parsed.query)
        if "v" in qs:
            return qs["v"][0]
        # /shorts/VIDEO_ID or /embed/VIDEO_ID
        parts = parsed.path.strip("/").split("/")
        if len(parts) >= 2 and parts[0] in ("shorts", "embed", "v"):
            return parts[1]
    return None


def _parse_duration(iso: str) -> Optional[int]:
    """Convert ISO 8601 duration (PT1H2M3S) to seconds."""
    m = re.match(r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?", iso)
    if not m:
        return None
    return (int(m.group(1) or 0) * 3600) + (int(m.group(2) or 0) * 60) + int(m.group(3) or 0)


class YouTubeAdapter(PlatformAdapter):
    @property
    def platform(self) -> SurfPlatform:
        return SurfPlatform.YOUTUBE

    def can_handle(self, url: str) -> bool:
        parsed = urlparse(url)
        return parsed.hostname in _YT_HOSTS

    # ── Search ────────────────────────────────────────────────────────

    async def search(
        self, query: str, *, limit: int = 5, subreddit: Optional[str] = None,
    ) -> tuple[list[SearchResult], Optional[str], SearchAccessMode]:
        settings = get_settings()
        yt_key = getattr(settings, "youtube_api_key", None)

        # Try YouTube Data API v3 if key available
        if yt_key:
            return await self._search_api(query, limit, yt_key)

        # Fallback: site-scoped web search
        return await self._search_web(query, limit)

    async def _search_api(
        self, query: str, limit: int, api_key: str,
    ) -> tuple[list[SearchResult], Optional[str], SearchAccessMode]:
        try:
            async with httpx.AsyncClient(timeout=15) as client:
                resp = await client.get(
                    "https://www.googleapis.com/youtube/v3/search",
                    params={
                        "part": "snippet",
                        "q": query,
                        "type": "video",
                        "maxResults": min(limit, 25),
                        "key": api_key,
                    },
                )
                if resp.status_code != 200:
                    return [], f"YouTube API {resp.status_code}", SearchAccessMode.UNAVAILABLE

                data = resp.json()
                results = []
                for item in data.get("items", [])[:limit]:
                    vid = item["id"].get("videoId", "")
                    snippet = item.get("snippet", {})
                    results.append(SearchResult(
                        url=f"https://www.youtube.com/watch?v={vid}",
                        title=snippet.get("title", ""),
                        snippet=snippet.get("description", "")[:300],
                        provider="youtube_api",
                        access_mode=SearchAccessMode.PUBLIC,
                        platform="youtube",
                        published_at=snippet.get("publishedAt", ""),
                    ))
                return results, None, SearchAccessMode.PUBLIC
        except Exception as e:
            return [], str(e), SearchAccessMode.UNAVAILABLE

    async def _search_web(
        self, query: str, limit: int,
    ) -> tuple[list[SearchResult], Optional[str], SearchAccessMode]:
        try:
            from ..search.engine import relai_search
            result = await relai_search(f"site:youtube.com {query}", limit=limit)
            hits = result.get("hits", [])
            filtered = [h for h in hits if any(
                host in h.get("url", "")
                for host in ("youtube.com", "youtu.be")
            )]
            if not filtered:
                return [], "No YouTube results", SearchAccessMode.PUBLIC
            return [
                SearchResult(
                    url=h["url"], title=h["title"], snippet=h.get("snippet", ""),
                    provider="youtube_web", access_mode=SearchAccessMode.PUBLIC,
                    platform="youtube",
                )
                for h in filtered
            ], None, SearchAccessMode.PUBLIC
        except Exception as e:
            return [], str(e), SearchAccessMode.UNAVAILABLE

    # ── Extraction ────────────────────────────────────────────────────

    async def extract_page(self, url: str) -> Optional[PageData]:
        vid = _video_id(url)
        if not vid:
            return None

        # Try oEmbed for structured metadata
        oembed = await self._oembed(vid)
        # Try page fetch for additional data
        page = await self._fetch_page(url)

        title = ""
        author = ""
        text_parts: list[str] = []

        if oembed:
            title = oembed.get("title", "")
            author = oembed.get("author_name", "")
            text_parts.append(f"# {title}")
            text_parts.append(f"Channel: {author}")
            if oembed.get("thumbnail_url"):
                text_parts.append(f"Thumbnail: {oembed['thumbnail_url']}")
            text_parts.append("")

        if page and page.text:
            # Extract useful sections from page text
            lines = page.text.split("\n")
            for line in lines:
                stripped = line.strip()
                # Grab description, chapters, metadata
                if any(kw in stripped.lower() for kw in (
                    "description", "chapter", "published", "views",
                    "likes", "subscribe", "about",
                )):
                    text_parts.append(stripped)
                if len("\n".join(text_parts)) > 8000:
                    break

        text = "\n".join(text_parts) if text_parts else f"Video: {vid}"
        metadata: dict[str, str] = {"video_id": vid}
        if author:
            metadata["channel"] = author
        if oembed:
            metadata["type"] = oembed.get("type", "video")

        return PageData(
            url=url,
            title=title or f"YouTube Video {vid}",
            text=text,
            markdown=text,
            links=[],
            headings=[],
            metadata=metadata,
            truncated=False,
            status=200,
            content_type="text/markdown",
            access_mode=SearchAccessMode.PUBLIC,
        )

    async def _oembed(self, vid: str) -> Optional[dict]:
        try:
            async with httpx.AsyncClient(timeout=10) as client:
                resp = await client.get(
                    "https://www.youtube.com/oembed",
                    params={"url": f"https://www.youtube.com/watch?v={vid}", "format": "json"},
                )
                if resp.status_code == 200:
                    return resp.json()
        except Exception:
            pass
        return None

    async def _fetch_page(self, url: str) -> Optional[PageData]:
        try:
            result = await relai_fetch(url, timeout_ms=12_000, retries=1)
            if not result.ok:
                return None
            from ..extraction.page_data import extract_page_data
            return extract_page_data(
                html=result.text, url=result.url, status=result.status,
                content_type=result.content_type, max_chars=15_000,
            )
        except Exception:
            return None
